stan/math/rev/mat/fun/cov_per.hpp

```diff
@@ -0,0 +1,347 @@
+						)
+        : vari(0.0),
+          size_(x.size()),
+          size_ltri_(size_ * (size_ - 1) / 2),
+          l_d_(value_of(l)), sigma_d_(value_of(sigma)),
+		  p_d_(value_of(p)),
+          sigma_sq_d_(sigma_d_ * sigma_d_),
+		  dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  cos_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_sq_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          l_vari_(l.vi_), sigma_vari_(sigma.vi_),
+		  p_vari_(p.vi_),
+          cov_lower_(ChainableStack::memalloc_.alloc_array<vari*>(size_ltri_)),
+          cov_diag_(ChainableStack::memalloc_.alloc_array<vari*>(size_)) {
+        double neg_two_inv_l_sq = -2.0 / (l_d_ * l_d_);
+        double pi_div_p = M_PI / p_d_;
+
+        size_t pos = 0;
+        for (size_t j = 0; j < size_ - 1; ++j) {
+          for (size_t i = j + 1; i < size_; ++i) {
+        	double dist = distance(x[i], x[j]);
+        	double sin_dist = sin(pi_div_p * dist);
+            double sin_dist_sq = square(sin_dist);
+            dist_[pos] = dist;
+            sin_dist_[pos] = sin_dist;
+        	cos_dist_[pos] = cos(pi_div_p * dist);
+            sin_dist_sq_[pos] = sin_dist_sq;
+            cov_lower_[pos] = new vari(sigma_sq_d_ * std::exp(sin_dist_sq
+                                       * neg_two_inv_l_sq), false);
+            ++pos;
+          }
+        }
+        for (size_t i = 0; i < size_; ++i)
+          cov_diag_[i] = new vari(sigma_sq_d_, false);
+      }
+
+      virtual void chain() {
+        double adjl = 0;
+        double adjsigma = 0;
+        double adjp = 0;
+
+        for (size_t i = 0; i < size_ltri_; ++i) {
+          vari* el_low = cov_lower_[i];
+          double prod_add = el_low->adj_ * el_low->val_;
+          adjl += prod_add * sin_dist_sq_[i];
+          adjsigma += prod_add;
+          adjp += prod_add * sin_dist_[i] * cos_dist_[i] * dist_[i];
+        }
+        for (size_t i = 0; i < size_; ++i) {
+          vari* el = cov_diag_[i];
+          adjsigma += el->adj_ * el->val_;
+        }
+        double l_d_sq = l_d_ * l_d_;
+        l_vari_->adj_ +=  adjl * 4 / (l_d_sq * l_d_);
+        sigma_vari_->adj_ += adjsigma * 2 / sigma_d_;
+        p_vari_->adj_ +=  adjp * 4 * M_PI / l_d_sq / (p_d_ * p_d_);
+      }
+    };
+
+    /**
+     * This is a subclass of the vari class for precomputed
+     * gradients of cov_per.
+     *
+     * The class stores the double values for the distance
+     * matrix, pointers to the varis for the covariance
+     * matrix, along with a pointer to the vari for sigma,
+     * the vari for l and the vari for p.
+     *
+     * @tparam T_x type of std::vector of elements
+     * @tparam T_l type of length scale
+     * @tparam T_p type of period
+     */
+    template <typename T_x, typename T_l, typename T_p>
+    class cov_per_vari<T_x, double, T_l, T_p> : public vari {
```

Comment:
This is hard for me to comment as I used cov_exp_quad as a template. For some reason there is a specialization there. Here we have of course more parameters and maybe it would be more important to actually have similar specialization for p instead. I would assume that use cases with fixed period would be more realistic.